### ISSUE TITLE
Unsafe SafeMaskMaker

### ISSUE THREAD

**Issue by bkhelifi** [gammapy/gammapy#3978]
**Gammapy version**
The dev version

**Bug description**
The issue happens during the Data Reduction step, without the use of a cutout. When using a large FoV of the final stacked `MapDataset`, the processing by the `SafeMaskMaker` of an observation with a pointing position far from the center of the final `MapDataset` (e.g. >2.5deg for HESS) leads to a python error. The `mask_safe` can not be computed (crash in L176 of `makers/safe.py`).
If a cutout centered on the pointing position is used to make an empty `MapDataset` from the final stacked `MapDataset`, the `SafeMaskMaker` works correctly.

It seems that there are two issues:
- the handling of the `position` parameter of `SafeMastMaker`: the observation center is not used for a standard treatment of a run (in `make_mask_energy_aeff_max` and also in `make_mask_energy_edisp_bias`)
- if no threshold is found in `make_mask_energy_aeff_max`, one should return an empty mask

**Expected behavior**
If an observation is at the very edge of a MapDataset, the worst case should be that an empty mask is return.
In all other case, no crash should happen and the mask_safe should be correctly computed.

**To Reproduce**
```
import astropy.units as u
from gammapy.maps import Map, MapAxis, WcsGeom
from gammapy.datasets import MapDataset
from gammapy.makers import MapDatasetMaker, SafeMaskMaker
data_store = DataStore.from_dir("$GAMMAPY_DATA/hess-dl3-dr1")
selection = dict(
    type="sky_circle",
    frame="icrs",
    lon="83.633 deg",
    lat="22.014 deg",
    radius="5 deg",
)
selected_obs_table = data_store.obs_table.select_observations(selection)
observations = data_store.get_observations(selected_obs_table["OBS_ID"])
energy_axis = MapAxis.from_energy_bounds(1.0, 10.0, 4, unit="TeV")

geom = WcsGeom.create(
    skydir=(85.633, 22.014),
    binsz=0.02,
    width=(10, 10),
    frame="icrs",
    proj="CAR",
    axes=[energy_axis],
)

# Reduced IRFs are defined in true energy (i.e. not measured energy).
energy_axis_true = MapAxis.from_energy_bounds(
    0.5, 20, 10, unit="TeV", name="energy_true"
)

stacked = MapDataset.create(
    geom=geom, energy_axis_true=energy_axis_true, name="crab-stacked"
)
offset_max = 2.5 * u.deg
maker = MapDatasetMaker()
maker_safe_mask = SafeMaskMaker(
    methods=["offset-max", "aeff-max"], offset_max=offset_max
)

for obs in observations:
    empty = stacked.copy(name=f"obs-{obs.obs_id}")
    # A MapDataset is filled in this cutout geometry
    dataset = maker.run(empty, obs)
    # The data quality cut is applied
    dataset = maker_safe_mask.run(dataset, obs)

```
**Other information**
Any other information you think will be useful for us to fix the issue can go here.

**Comment by adonath** [comment]
Yes, for ID 23592 the offset is actually larger than 2.5 deg and thus the effective area is zero. That's why the option "aeff-max" does not work. I'm not sure what the expected behavior should be in this case. One could go for the nearest valid value for effective area or just ask user to use the `fixed_offset` option in this case?

**Comment by AtreyeeS** [comment]
Probably the centre of the map is not a good default choice? If an observation is passed, the pointing position (or a fixed offset from it) should be used. Else, the centre of the geom.
In the latter case, probably returning a mask with all False is a valid option.

**Comment by registerrier** [comment]
Thanks @bkhelifi . 

The first issue you mention regarding the fact that `SafeMaskMaker` will extract the aeff at the `geom` center rather than at the pointing position. This is actually consistent with what the `SafeMaskMaker` documentation says. 
Obviously, this is a subtle point in the way we document the code and write the tutorials. As noted by @AtreyeeS there are two possible defaults, center of the `Dataset` `geom` or `offset=0`. They coincide most of the time only because we perform a cutout all the time. 
Maybe changing the default behavior depending on the arguments passed (i.e. whether `obs=None` or not) as suggested by @AtreyeeS is even harder to properly document.

For the second issue, rather than failing returning a `False` map seems like the best option.

**Comment by registerrier** [comment]
BTW, I think there is an issue in the `SafeMaskMaker` documentation: `position` will take precedence over `fixed_offset` but this is not clearly stated. 
I assume a safe behavior would be to make the `Maker` fail if both arguments are set at the same time.

**Comment by bkhelifi** [comment]
@registerrier (2nd comment): I agree, even if it seems that in this exemple we are not in this case
@AtreyeeS and @adonath : when having a very large RoI, its center has not more meaning for an individual observation. An observation should not be masked if it is far from this center, it should be used according to the asked criteria. The pointing position is then the only reference and then the option `fixed_offset` is somehow mandatory with a value of 0 deg.
To all (associated to the first comment of @registerrier ): one still has to return a fully masked dataset instead of a Python error.

If I do not mistaken, we have 2 main use cases:
1. point-source analysis for wobble observations: the observations are close to the map center and the latter can be used (the safe mask is then well computed as the map center option is similar to the `fixed_offset` option)
2. RoI analysis: the observations are in general far from the map center 
For 2., fixed_offset=0deg will do the job, or position=obs.pointing_position. For 1, either nothing or position=obs.pointing_position. Do I mistaken?
Is there other use cases?

**Comment by AtreyeeS** [comment]
@bkhelifi for the case what you term "RoI analysis", I think we **always** encourage users to work with cutouts, ie, inside the data reduction loop, use

`empty = stacked.cutout(obs.pointing_radec, width=2 * offset_max, name=f"obs-{obs.obs_id}" )`
instead of `empty = stacked.copy(name=f"obs-{obs.obs_id}")`
In this case, the centre of `empty` is indeed the pointing position, and the problem resolves itself.

It may be an option for gammapy to internally make the cutouts, and I believe this is already implemented in the `DatasetsMaker`

**Comment by adonath** [comment]
> I assume a safe behavior would be to make the Maker fail if both arguments are set at the same time.

This is already the case: https://github.com/gammapy/gammapy/blob/master/gammapy/makers/safe.py#L75

**Comment by adonath** [comment]
Addressed in #3987.

**Comment by bkhelifi** [comment]
With the current code, the example returns:

`Effective area is all zero at [85d37m58.8s 22d00m50.4s]. No safe energy band can be defined for the dataset 'obs-23592': setting `mask_safe` to all False.
`
So, I think that we are safe now!

A second point is the improvement of the documentation about the Data Reduction as mentioned @AtreyeeS . One should use:
`empty = stacked.cutout(obs.pointing_radec, width=2 * offset_max, name=f"obs-{obs.obs_id}" )
`
instead of 
`empty = stacked.copy(name=f"obs-{obs.obs_id}")
`
I think that this is well described in this tutorial [https://docs.gammapy.org/0.20.1/tutorials/starting/analysis_2.html](https://docs.gammapy.org/0.20.1/tutorials/starting/analysis_2.html).

So, I propose to close this issue.
